keras_nlp/models/deberta/disentangled_self_attention.py

```diff
@@ -0,0 +1,376 @@
+            ):
+                attention_mask = tf.expand_dims(
+                    attention_mask, axis=mask_expansion_axis
+                )
+        return self._softmax(attention_scores, attention_mask)
+
+    def _compute_attention(
+        self,
+        query,
+        key,
+        value,
+        rel_embeddings,
+        attention_mask=None,
+        training=None,
+    ):
+        """Computes the attention score and returns the attended outputs.
+
+        This function computes vanilla MHA score, and relative attention scores
+        (p2c and c2p). It then sums them up to get the final attention score,
+        which is used to compute the attended outputs.
+        """
+
+        attention_scores = tf.einsum(
+            "aecd,abcd->acbe",
+            key,
+            query,
+        )
+        attention_scores = tf.multiply(attention_scores, self.scale_factor)
+
+        rel_embeddings = self._position_dropout_layer(
+            rel_embeddings,
+            training=training,
+        )
+
+        rel_attn_scores = self._compute_disentangled_attention(
+            query=query,
+            key=key,
+            rel_embeddings=rel_embeddings,
+        )
+
+        if rel_attn_scores is not None:
+            attention_scores += rel_attn_scores
+
+        attention_scores = self._masked_softmax(
+            attention_scores, attention_mask
+        )
+        attention_scores = self._attn_dropout_layer(
+            attention_scores, training=training
+        )
+        attention_output = tf.einsum("acbe,aecd->abcd", attention_scores, value)
+
+        return attention_output, attention_scores
+
+    def _make_log_bucket_position(self, rel_pos):
+        dtype = rel_pos.dtype
+        sign = tf.math.sign(rel_pos)
+        mid = self.bucket_size // 2
+        mid = tf.cast(mid, dtype=dtype)
+
+        # If `rel_pos[i][j]` is out of bounds, assign value `mid`.
+        abs_pos = tf.where(
+            condition=(rel_pos < mid) & (rel_pos > -mid),
+            x=mid - 1,
+            y=tf.math.abs(rel_pos),
+        )
+
+        def _get_log_pos(abs_pos, mid):
+            numerator = tf.math.log(abs_pos / mid)
+            numerator = numerator * tf.cast(mid - 1, dtype=numerator.dtype)
+            denominator = tf.math.log((self.max_position_embeddings - 1) / mid)
+            val = tf.math.ceil(numerator / denominator)
+            val = tf.cast(val, dtype=mid.dtype)
+            val = val + mid
+            return val
+
+        log_pos = _get_log_pos(abs_pos, mid)
+
+        bucket_pos = tf.where(
+            condition=abs_pos <= mid,
+            x=rel_pos,
+            y=log_pos * sign,
+        )
+        bucket_pos = tf.cast(bucket_pos, dtype=tf.int64)
+
+        return bucket_pos
+
+    def _get_rel_pos(self, num_positions):
+        ids = tf.range(num_positions, dtype=tf.int64)
+        query_ids = ids[:, tf.newaxis]
+        # query_ids = tf.repeat(query_ids, repeats=num_positions, axis=1)
```

Comment:
why is this line commented out? remove?

Reply:
Yeah, the broadcasting stuff is done by Keras on its own. So, `tf.repeat` is not needed.